pkg/reconciler/pipelinerun/resources/apply_test.go

```diff
@@ -2034,6 +2034,284 @@ func TestApplyParameters_ArrayIndexing(t *testing.T) {
+				Name: "first-param", Type: v1beta1.ParamTypeString, Default: v1beta1.NewStructuredValues("default-value"),
+			}},
+			Tasks: []v1beta1.PipelineTask{{
+				Matrix: &v1beta1.Matrix{
+					Include: []v1beta1.MatrixInclude{{
+						Name: "build-1",
+						Params: v1beta1.Params{{
+							Name: "first-param", Value: *v1beta1.NewStructuredValues("default-value"),
+						}}},
+					}},
+			}},
+		},
+	}, {
+		name: "matrix include params - string replacement - override default",
+		original: v1beta1.PipelineSpec{
+			Params: []v1beta1.ParamSpec{{
+				Name: "first-param", Type: v1beta1.ParamTypeString, Default: v1beta1.NewStructuredValues("default-value"),
+			}, {
+				Name: "second-param", Type: v1beta1.ParamTypeString}},
+			Tasks: []v1beta1.PipelineTask{{
+				Matrix: &v1beta1.Matrix{
+					Include: []v1beta1.MatrixInclude{{
+						Name: "build-1",
+						Params: v1beta1.Params{{
+							Name: "first-param", Value: *v1beta1.NewStructuredValues("$(params.first-param)"),
+						}, {
+							Name: "second-param", Value: *v1beta1.NewStructuredValues("$(params.second-param)"),
+						}, {
+							Name: "third-param", Value: *v1beta1.NewStructuredValues("new-value"),
+						}}},
+					}},
+			}},
+		},
+		params: v1beta1.Params{{Name: "second-param", Value: *v1beta1.NewStructuredValues("second-value")}},
+		expected: v1beta1.PipelineSpec{
+			Params: []v1beta1.ParamSpec{{
+				Name: "first-param", Type: v1beta1.ParamTypeString, Default: v1beta1.NewStructuredValues("default-value"),
+			}, {
+				Name: "second-param", Type: v1beta1.ParamTypeString,
+			}},
+			Tasks: []v1beta1.PipelineTask{{
+				Matrix: &v1beta1.Matrix{
+					Include: []v1beta1.MatrixInclude{{
+						Name: "build-1",
+						Params: v1beta1.Params{{
+							Name: "first-param", Value: *v1beta1.NewStructuredValues("default-value"),
+						}, {
+							Name: "second-param", Value: *v1beta1.NewStructuredValues("second-value"),
+						}, {
+							Name: "third-param", Value: *v1beta1.NewStructuredValues("new-value"),
+						}}},
+					}},
+			}},
+		},
+	}, {name: "matrix include params - string replacement from object params",
+		original: v1beta1.PipelineSpec{
+			Params: []v1beta1.ParamSpec{{
+				Name: "myobject",
+				Properties: map[string]v1beta1.PropertySpec{
+					"key1": {Type: "string"},
+					"key2": {Type: "string"},
+				},
+				Default: v1beta1.NewObject(map[string]string{
+					"key1": "default",
+					"key2": "param!",
+				}),
+			}},
+			Tasks: []v1beta1.PipelineTask{{
+				Matrix: &v1beta1.Matrix{
+					Include: []v1beta1.MatrixInclude{{
+						Name: "build-1",
+						Params: v1beta1.Params{{
+							Name: "first-param", Value: *v1beta1.NewStructuredValues("$(params.myobject.key1)"),
+						}, {
+							Name: "second-param", Value: *v1beta1.NewStructuredValues("$(params.myobject.key2)"),
+						}}},
+					}},
+			}},
+		},
+		params: v1beta1.Params{{Name: "myobject", Value: *v1beta1.NewObject(map[string]string{
+			"key1": "pipeline",
+			"key2": "param!!",
```

Comment:
this value is `param!!` and the default value is `param!` ? Wonder if we could use some values easier to tell the differences?